python/ray/tune/utils/resource_updater.py

```diff
@@ -0,0 +1,133 @@
+from typing import Optional, Dict, Any
+import logging
+import os
+import time
+
+import ray
+from ray import ray_constants
+from ray._private.resource_spec import NODE_ID_PREFIX
+from ray.tune.resources import Resources
+
+logger = logging.getLogger(__name__)
+
+TUNE_STATE_REFRESH_PERIOD = 10  # Refresh resources every 10 s
+
+
+def _to_gb(n_bytes):
+    return round(n_bytes / (1024 ** 3), 2)
+
+
+class ResourceUpdater:
+    """Periodic Resource updater for Tune."""
+
+    def __init__(self, refresh_period: Optional[float] = None):
+        self._avail_resources = Resources(cpu=0, gpu=0)
+        self._resources_initialized = False
+
+        if refresh_period is None:
+            refresh_period = float(
+                os.environ.get("TUNE_STATE_REFRESH_PERIOD", TUNE_STATE_REFRESH_PERIOD)
+            )
+        self._refresh_period = refresh_period
+        self._last_resource_refresh = float("-inf")
+
+        if ray.is_initialized():
+            self.update_avail_resources()
+
+    def update_avail_resources(self, num_retries=5):
+        if time.time() - self._last_resource_refresh < self._refresh_period:
+            return
+        logger.debug("Checking Ray cluster resources.")
+        resources = None
+        for i in range(num_retries):
+            if i > 0:
+                logger.warning(
+                    "Cluster resources not detected or are 0. Attempt #" "%s...", i + 1
+                )
+                time.sleep(0.5)
+            resources = ray.cluster_resources()
+            if resources:
+                break
+
+        if not resources:
+            # NOTE: This hides the possibility that Ray may be waiting for
+            # clients to connect.
+            resources.setdefault("CPU", 0)
+            resources.setdefault("GPU", 0)
+            logger.warning(
+                "Cluster resources cannot be detected or are 0. "
+                "You can resume this experiment by passing in "
+                "`resume=True` to `run`."
+            )
+
+        resources = resources.copy()
+        num_cpus = resources.pop("CPU", 0)
+        num_gpus = resources.pop("GPU", 0)
+        memory = ray_constants.from_memory_units(resources.pop("memory", 0))
+        object_store_memory = ray_constants.from_memory_units(
+            resources.pop("object_store_memory", 0)
+        )
+        custom_resources = resources
+
+        self._avail_resources = Resources(
+            int(num_cpus),
+            int(num_gpus),
+            memory=int(memory),
+            object_store_memory=int(object_store_memory),
+            custom_resources=custom_resources,
+        )
+        self._last_resource_refresh = time.time()
+        self._resources_initialized = True
```

Comment:
I wonder if this variable could be covered by `_last_resource_refresh` > 0.

Reply:
I think so. good idea